src/server/pps/server/pipeline_controller_test.go

```diff
@@ -0,0 +1,686 @@
+		{
+			pipeline:   pipeline,
+			assertWhen: []pps.PipelineState{pps.PipelineState_PIPELINE_RUNNING},
+			expectedStates: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_STARTING,
+				pps.PipelineState_PIPELINE_RUNNING,
+			},
+		},
+	})
+	require.Equal(t, 1, infraDriver.calls[pipeline][mockInfraOp_CREATE])
+	// clear state and trigger the delete event
+	stateDriver.reset()
+	defer stateDriver.cancelWatch()
+	stateDriver.pushWatchEvent(pi, watch.EventDelete)
+	require.NoErrorWithinT(t, 5*time.Second, func() error {
+		return backoff.Retry(func() error {
+			if infraDriver.calls[pipeline][mockInfraOp_DELETE] == 1 {
+				return nil
+			}
+			return errors.New("change hasn't reflected")
+		}, backoff.NewTestingBackOff())
+	})
+}
+
+func TestDeleteRC(t *testing.T) {
+	stateDriver, infraDriver, _ := ppsMasterHandles(t)
+	pipeline := tu.UniqueString(t.Name())
+	pi := &pps.PipelineInfo{
+		Pipeline: client.NewPipeline(pipeline),
+		State:    pps.PipelineState_PIPELINE_STARTING,
+		Details:  &pps.PipelineInfo_Details{},
+		Version:  1,
+	}
+	stateDriver.upsertPipeline(pi)
+	validate(t, stateDriver, infraDriver, []pipelineTest{
+		{
+			pipeline:   pipeline,
+			assertWhen: []pps.PipelineState{pps.PipelineState_PIPELINE_RUNNING},
+			expectedStates: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_STARTING,
+				pps.PipelineState_PIPELINE_RUNNING,
+			},
+		},
+	})
+	require.Equal(t, 1, infraDriver.calls[pipeline][mockInfraOp_CREATE])
+	// remove RCs, and nudge with an event
+	infraDriver.resetRCs()
+	stateDriver.pushWatchEvent(pi, watch.EventPut)
+	// verify restart side effects were requested
+	require.NoErrorWithinT(t, 5*time.Second, func() error {
+		return backoff.Retry(func() error {
+			if infraDriver.calls[pipeline][mockInfraOp_CREATE] == 2 {
+				return nil
+			}
+			return errors.New("change hasn't reflected")
+		}, backoff.NewTestingBackOff())
+	})
+
+}
+
+func TestAutoscalingBasic(t *testing.T) {
+	stateDriver, infraDriver, mockPachd := ppsMasterHandles(t)
+	pipeline := tu.UniqueString(t.Name())
+	mockAutoscaling(mockPachd, 1, 1)
+	stateDriver.upsertPipeline(&pps.PipelineInfo{
+		Pipeline: client.NewPipeline(pipeline),
+		State:    pps.PipelineState_PIPELINE_STARTING,
+		Details: &pps.PipelineInfo_Details{
+			Autoscaling: true,
+			ParallelismSpec: &pps.ParallelismSpec{
+				Constant: 1,
+			},
+		},
+		Version: 1,
+	})
+	validate(t, stateDriver, infraDriver, []pipelineTest{
+		{
+			pipeline: pipeline,
+			assertWhen: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_RUNNING,
+			},
+			expectedStates: []pps.PipelineState{
+				pps.PipelineState_PIPELINE_STARTING,
+				// STANDBY is set consecutively because ppsutil.SetPipelineState
+				// will still update the PipelineInfo even if the desired state
+				// is already set
+				pps.PipelineState_PIPELINE_STANDBY,
+				pps.PipelineState_PIPELINE_STANDBY,
+				pps.PipelineState_PIPELINE_RUNNING,
+				pps.PipelineState_PIPELINE_STANDBY,
```

Comment:
ok, I'm definitely missing something - isn't the `assertWhen` going to succeed when the last state is RUNNING? Is it just going back quickly enough that STANDBY shows up anyway?

Reply:
yeah you're right. I removed `assertWhen` due to discussion above